plugins/deployment_manager/Spider.ts

```diff
@@ -1,211 +1,265 @@
+    if (maybeExists === address) {
+      return false;
+    } else {
+      throw new Error(`Had ${alias} -> ${maybeExists}, not ${address}`);
+    }
+  } else {
+    into.set(alias, address);
+    return true;
+  }
+}
+
+async function discoverNodes(
+  path: Contract[],
   contract: Contract,
-  relationConfig: RelationInnerConfig,
-  alias: Alias
-): Promise<[Address, AliasTemplate][]> {
-  let addresses: string[] = await readField(contract, getFieldKey(alias, relationConfig));
-  let aliasTemplates: AliasTemplate[] = relationConfig.alias
-    ? asArray<AliasTemplate>(relationConfig.alias)
-    : [aliasTemplateFromAlias(alias)];
-
-  return cross(addresses, aliasTemplates);
+  context: Ctx,
+  config: RelationInnerConfig,
+  defaultKeyAndTemplate: string
+): Promise<DiscoverNode[]> {
+  const addresses = await readField(contract, getFieldKey(config, defaultKeyAndTemplate), context);
+  const templates = config.alias ? asArray(config.alias) : [defaultKeyAndTemplate];
+  return addresses.map((address, i) => ({
+    address,
+    aliasRender: { template: templates[i % templates.length], i },
+    path: [contract].concat(path),
+  }));
 }
 
-interface VisitedNode {
-  buildFile: BuildFile;
-  contract: Contract;
-  aliasTemplates: AliasTemplate[];
-  implAddress?: Address;
+async function isContract(hre: HRE, address: string) {
+  return await hre.ethers.provider.getCode(address) !== '0x';
+}
+
+async function localBuild(cache: Cache, hre: HRE, artifact: string, address: Address): Promise<Build> {
+  const buildFile = cache ?
+    await readAndCacheContract(cache, hre, artifact, address) :
+    await readContract(null, hre, artifact, address, true);
+  const contract = getEthersContract(address, buildFile, hre);
+  return { buildFile, contract };
 }
 
-// Tail-call optimized version of spider method, which crawls a dependency graph gathering contract data
-async function runSpider(
+async function remoteBuild(cache: Cache, hre: HRE, network: string, address: Address): Promise<Build> {
+  const buildFile = await fetchAndCacheContract(cache, network, address);
+  const contract = getEthersContract(address, buildFile, hre);
+  return { buildFile, contract };
+}
+
+async function crawl(
   cache: Cache,
   network: string,
-  hre: HardhatRuntimeEnvironment,
-  relationConfigMap: RelationConfigMap, // For base relations (??)
-  discovered: [Address, AliasTemplate][],
-  visited: Map<Address, VisitedNode>,
-  importRetries?: number,
-  importRetryDelay?: number
-): Promise<Map<Address, VisitedNode>> {
-  if (discovered.length === 0) {
-    // We have no more unvisited nodes; that means we're done.
-    return visited;
-  }
+  hre: HRE,
+  relations: RelationConfigMap,
+  node: DiscoverNode,
+  context: Ctx,
+  aliases: Aliases,
+  proxies: Proxies,
+  contracts: ContractMap,
+): Promise<Alias> {
+  const { aliasRender, address, path } = node;
+  const { template: aliasTemplate } = aliasRender;
+  //debug(`Crawling ${address}`, aliasRender);
```

Comment:
Uncomment all these `debug`s if you are trying to follow what's happening for some reason and it'll all be super clear